stdlib/universe/moving_average.go

```diff
@@ -0,0 +1,557 @@
+		if cr.Len() == 0 {
+			return nil
+		}
+
+		for j, c := range cr.Cols() {
+			var err error
+			switch c.Type {
+			case flux.TBool:
+				err = t.passThroughBool(cr.Bools(j), builder, j)
+			case flux.TInt:
+				err = t.doInt(cr.Ints(j), builder, j, doMovingAverage[j])
+			case flux.TUInt:
+				err = t.doUInt(cr.UInts(j), builder, j, doMovingAverage[j])
+			case flux.TFloat:
+				err = t.doFloat(cr.Floats(j), builder, j, doMovingAverage[j])
+			case flux.TString:
+				err = t.passThroughString(cr.Strings(j), builder, j)
+			case flux.TTime:
+				err = t.passThroughTime(cr.Times(j), builder, j)
+			}
+
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+func (t *movingAverageTransformation) UpdateWatermark(id execute.DatasetID, mark execute.Time) error {
+	return t.d.UpdateWatermark(mark)
+}
+
+func (t *movingAverageTransformation) UpdateProcessingTime(id execute.DatasetID, pt execute.Time) error {
+	return t.d.UpdateProcessingTime(pt)
+}
+
+func (t *movingAverageTransformation) Finish(id execute.DatasetID, err error) {
+	t.d.Finish(err)
+}
+
+func (t *movingAverageTransformation) passThroughBool(vs *array.Boolean, b execute.TableBuilder, bj int) error {
+	if int64(vs.Len()) < t.n {
+		if vs.IsNull(vs.Len() - 1) {
+			if err := b.AppendNil(bj); err != nil {
+				return err
+			}
+		} else {
+			if err := b.AppendBool(bj, vs.Value(vs.Len()-1)); err != nil {
+				return nil
+			}
+		}
+		return nil
+	} else {
+		s := arrow.BoolSlice(vs, int(t.n-1), vs.Len())
+		defer s.Release()
+		return b.AppendBools(bj, s)
+	}
+}
+
+func (t *movingAverageTransformation) doInt(vs *array.Int64, b execute.TableBuilder, bj int, doMovingAverage bool) error {
+	if t.window[bj] == nil {
+		t.window[bj] = make([]interface{}, t.n)
+	}
+	if t.sum[bj] == nil {
+		t.sum[bj] = new(int64)
+	}
+	sumPointer := &t.sum[bj]
+	sum := (*sumPointer).(*int64)
+
+	j := 0
+
+	if vs.IsValid(j) {
+		*sum += vs.Value(j)
+		t.count[bj]++
+		t.window[bj][0] = vs.Value(j)
+	} else {
+		t.window[bj][0] = nil
+	}
+	j++
+	t.i[bj]++
+
+	l := vs.Len()
+	for ; j < l; j++ {
+
+		if !vs.IsNull(j) {
+			t.count[bj]++
+			t.window[bj][int64(t.i[bj])%t.n] = vs.Value(j)
+		} else {
+			t.window[bj][int64(t.i[bj])%t.n] = nil
+		}
+		*sum += vs.Value(j)
+
+		if int64(t.i[bj]) < t.n-1 {
+			t.i[bj]++
+			continue
+		}
+
+		if !doMovingAverage {
+			if vs.IsValid(j) {
+				if err := b.AppendInt(bj, vs.Value(j)); err != nil {
+					return err
+				}
+			} else {
+				if err := b.AppendNil(bj); err != nil {
+					return err
+				}
+			}
+		} else {
+			average := 0.0
+			if t.count[bj] != 0 {
+				average = float64(*sum) / float64(t.count[bj])
+				if err := b.AppendFloat(bj, average); err != nil {
+					return err
+				}
+			} else {
+				if err := b.AppendNil(bj); err != nil {
+					return err
+				}
+			}
+
+			next := t.window[bj][int64(t.i[bj]+1)%t.n]
+			if next != nil {
+				*sum -= next.(int64)
+				t.count[bj]--
+			}
+
+		}
+		t.i[bj]++
+	}
+	if int64(t.i[bj]) < t.n-1 {
```

Comment:
This produces unwanted behavior; it should be called in Process after the RowReader. #1496